internal/db/disk/stats.go

```diff
@@ -8,86 +8,64 @@ import (
 	"github.com/edulinq/autograder/internal/util"
 )
 
-const SYSTEM_STATS_FILENAME = "stats.jsonl"
-const COURSE_STATS_FILENAME = "course-stats.jsonl"
-const API_REQUEST_STATS_FILENAME = "api-request-stats.jsonl"
+const SYSTEM_STATS_FILENAME = "system-stats.jsonl"
 
 func (this *backend) GetSystemStats(query stats.Query) ([]*stats.SystemMetrics, error) {
 	path := this.getSystemStatsPath()
 
-	this.statsLock.RLock()
-	defer this.statsLock.RUnlock()
+	this.systemStatsLock.RLock()
+	defer this.systemStatsLock.RUnlock()
 
 	records, err := util.FilterJSONLFile(path, stats.SystemMetrics{}, func(record *stats.SystemMetrics) bool {
-		return query.Match(record)
+		return query.MatchTimeWindow(record)
 	})
 
 	return records, err
 }
 
 func (this *backend) StoreSystemStats(record *stats.SystemMetrics) error {
-	this.statsLock.Lock()
-	defer this.statsLock.Unlock()
+	this.systemStatsLock.Lock()
+	defer this.systemStatsLock.Unlock()
 
 	return util.AppendJSONLFile(this.getSystemStatsPath(), record)
 }
 
-func (this *backend) GetCourseMetrics(query stats.CourseMetricQuery) ([]*stats.CourseMetric, error) {
-	if query.CourseID == "" {
-		return nil, fmt.Errorf("When querying for course metrics, course ID must not be empty.")
+func (this *backend) GetMetrics(query stats.Query) ([]*stats.Metric, error) {
+	path, err := this.getStatsPath(query.Type)
+	if err != nil {
+		return nil, err
 	}
 
-	path := this.getCourseStatsPath(query.CourseID)
-
-	this.contextReadLock(path)
-	defer this.contextReadUnlock(path)
+	this.statsLock.Lock()
+	defer this.statsLock.Unlock()
```

Comment:
I think you are misunderstanding the purpose of this lock,
it is being used incorrectly in two different ways.

1) This should be a read locking.
The idea behind this lock is that it signals to the backend that an operation is in progress and the backend can't close until we are done (hence the write Lock() in Close()).

2) You need to acquire a path-specific lock.
Since this lock is not about protecting the resources (it is about just telling the backend an operation is happening), you need another lock to protect the file you are working with.
More specifically, you need a lock for this file specifically (i.e., a contextual lock from the lockmanager).